### PR TITLE
app_memory: cleanup and fixed issues

### PR DESCRIPTION
**Pull request by AdithyaBaglody**
Gist of the changes. 

1. Added new macros. Which makes this feature more feasible to real use cases.
2.  Optimized calculation of partition sizes. Now all calculation are given by linker.
3. app_smem.ld is used by default on all architectures that support userspace.
4.  Fixed a few bugs in the code.
5.  Data/bss of libc comes in as a partition, which can be configured like any other partitions.
6. Updated the test and sample for app_smem.

cleanup patches taken from the PR. #11896. The test cases and use cases are still in the other PR.

### CODE REVIEW

**Review by SebastianBoe — DISMISSED** [review]

File: scripts/gen_app_partitions.py [hunk 1]
```diff
@@ -67,17 +83,29 @@ def find_partitions(filename, full_list_of_partitions, partitions_source_file):
...
-            full_list_of_partitions.remove(partition+"b")
+        if (partition+"_bss" in full_list_of_partitions):
+            full_list_of_partitions.remove(partition+"_bss")
     return full_list_of_partitions
 
-def generate_final_linker(linker_file, full_list_of_partitions):
+def libc_is_enabled(build_config_file):
+    with open(build_config_file, "r") as config_file_pointer:
+        full_file = config_file_pointer.readlines()
+        return "CONFIG_NEWLIB_LIBC=y\n" in full_file
```
[the remark this answers is not in the document]
Comment (AdithyaBaglody) [reply scripts/gen_app_partitions.py]: nice input. I didn't think of the cmake. Thanks for the input. @SebastianBoe please re check.

[the remark this answers is not in the document]
Comment (SebastianBoe) [reply scripts/gen_app_partitions.py]: Much better :)

File: include/arch/arm/cortex_m/scripts/linker.ld [hunk 2]
```diff
@@ -411,22 +411,9 @@ SECTIONS
...
-	{
-		APP_SHARED_ALIGN;
-		_app_smem_start = .;
-		APP_SMEM_SECTION()
-		/* Align the end of shared app mem section with the
-		 * minimum granularity required by MPU.
-		 */
-		APP_SHARED_ALIGN;
-		_app_smem_end = .;
+#define SMEM_PARTITION_ALIGN MPU_ALIGN
```
[the remark this answers is not in the document]
Comment (AdithyaBaglody) [reply include/arch/arm/cortex_m/scripts/linker.ld]: oh that's because the generated ld script is common for all architectures. it will be confusing if I mentioned MPU in x86 therefore a macro which is agnostic is being used.

File: scripts/gen_app_partitions.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (AdithyaBaglody) [reply scripts/gen_app_partitions.py]: @SebastianBoe can we have the -1 removed now?

[the remark this answers is not in the document]
Comment (SebastianBoe) [reply scripts/gen_app_partitions.py]: -1 removed

**Review by ruuddw — APPROVED** [review]
Works for me (tested on ARC with both types of MPU: power-of-two and arbitrary size regions).

File: include/arch/arc/v2/linker.ld [hunk 3]
```diff
@@ -139,22 +139,12 @@ SECTIONS {
...
-		APP_SMEM_SECTION()
-		MPU_MIN_SIZE_ALIGN
-		_app_smem_end = .;
-	} GROUP_DATA_LINK_IN(RAMABLE_REGION, ROMABLE_REGION)
+#define SMEM_PARTITION_ALIGN MPU_ALIGN
 
-#endif  /* CONFIG_MPU_REQUIRES_POWER_OF_TWO_ALIGNMENT */
+#include <app_smem.ld>
 
+	_image_ram_start = _app_smem_start;
```
Comment (ioannisg) [inline include/arch/arc/v2/linker.ld]: Could _image_ram_start be defined as in the ARM linker script?

File: tests/kernel/mem_protect/userspace/src/main.c [hunk 4]
```diff
@@ -38,14 +38,23 @@ K_SEM_DEFINE(expect_fault_sem, 0, 1);
  * ztest and this test suite. part1 and part2 are for
  * subsequent test specifically for this new implementation.
  */
-FOR_EACH(appmem_partition, part0, part1, part2);
+APPMEM_PARTITION_OBJECT_DEFINE(part0, part1, part2);
 
+/* this defines all the functions that are needed to configure
+ * The memory partition. Usually goes in a common header file.
```
Comment (ioannisg) [inline tests/kernel/mem_protect/userspace/src/main.c]: the

File: tests/kernel/mem_protect/userspace/src/main.c [hunk 5]
```diff
@@ -38,14 +38,23 @@ K_SEM_DEFINE(expect_fault_sem, 0, 1);
...
 /*
  * Create memory domains. dom0 is for the ztest and this
  * test suite, specifically. dom1 is for a specific test
  * in this test suite.
  */
-FOR_EACH(appmem_domain, dom0, dom1);
+APPMEM_DOMAIN_OBJECT_DEFINE(dom0, dom1);
+
+/* this defines all the functions that are needed to configure
+ * The memory domains. Usually goes in a common header file.
```
Comment (ioannisg) [inline tests/kernel/mem_protect/userspace/src/main.c]: the

File: tests/kernel/mem_protect/userspace/src/main.c [hunk 6]
```diff
@@ -688,23 +703,13 @@ extern k_thread_stack_t ztest_thread_stack[];
 void test_main(void)
 {
 	/* partitions must be initialized first */
-	FOR_EACH(appmem_init_part, part0, part1, part2);
+	APPMEM_INIT_PARTITIONS(part0, part1, part2);
 	/*
 	 * Next, the app_memory must be initialized in order to
 	 * calculate size of the dynamically created subsections.
 	 */
-#if defined(CONFIG_ARC)
```
Comment (ioannisg) [inline tests/kernel/mem_protect/userspace/src/main.c]: This change looks "un-related" to the pull-request, IMHO. This should be submitted separately and get ARC guys review it and confirm that such patches are not needed; i.e. supervisors have access to configure and initialize memory without need for disabling and re-enabling mpu...

File: subsys/app_memory/app_memdomain.c [hunk 7]
```diff
@@ -25,73 +25,26 @@ void app_bss_zero(void)
...
-}
-
 /*
  * "Initializes" by calculating subsection sizes and then
  * zeroizing "bss" regions.
  */
 void appmem_init_app_memory(void)
 {
-	app_calc_size();
+#ifdef CONFIG_ARC
```
Comment (ioannisg) [inline subsys/app_memory/app_memdomain.c]: Same comment as in the test source file below - this is not really related to the PR - this is something we need to think of and address explicitly for ARC. It relates to the MPU configuration.

**Review by ioannisg — CHANGES_REQUESTED** [review]
Some minor picks, and a comment about the ARC change
